docker/webapp/app.py

```diff
@@ -0,0 +1,258 @@
+# TODO:
+#  - rss
+#  - opensearch
+#  - unify static files, no dup melpa.css etc
+#  - toggle stable vs. non-stable, add link at top-right to the other
+#  - build timestamps
+#  - page titles
+#  - entity-quote package URLs containing +
+#  - case-insensitive search
+#  - full URLs for badges
+#  - gist recipes
+#  - escaping chars in description etc.
+#  - serve packages statically by default, use /packages/foo URLs.
+
+from fastapi import FastAPI, Request, HTTPException
+from fastapi.exceptions import RequestValidationError
+from fastapi.responses import HTMLResponse, PlainTextResponse, RedirectResponse
+from fastapi.staticfiles import StaticFiles
+from fastapi.templating import Jinja2Templates
+from starlette.datastructures import URL
+import starlette.exceptions
+import json
+import os
+import datetime
+from collections import namedtuple
+import re
+from itertools import groupby
+from math import ceil
+
+##############################################################################
+# Data models
+##############################################################################
+
+HTML_DIR="../../html"
+
+BuildStatus = namedtuple("BuildStatus", ['started_at', 'completed_at', 'next_at', 'duration'])
```

Comment:
Consider using [`@dataclass`](https://docs.python.org/3/library/dataclasses.html) if there's not a significant performance increase by having `namedtuple()`, it's more readable and allows specifying [types](https://docs.python.org/3/library/typing.html) and allow better static checking with [mypy](https://github.com/python/mypy).

Reply:
Just aiming for the most lightweight and concise object-instead-of-a-dict wrapper tbh, but I might consider dataclasses.